### ISSUE TITLE
Mailu Front Serving Redirect Loop After Scaling Nodes

### ISSUE THREAD

**Issue by Matthew-Benson** [Mailu/Mailu#1657]
## Environment
Kubernetes bare metal cluster with NFS volumes from local FreeNAS server.
CNI: Calico
Letsencrypt: Cert-manager
Ingress: Traefik v2
Volume Provisioner: nfs-client-provisioner

## Data Flow
Router > Traefik load balencer service > Mailu front pod > Mailu services

## Versions
Mailu: master
Kubernetes: v1.18.6
Calico: v3.15.1
nfs-client-provisioner: v3.1.0
Traefik: v2.2.8

## Config
dovecot.conf
```conf
mail_nfs_index = yes
mail_nfs_storage = yes
mail_fsync = always
mmap_disable = yes
mail_max_userip_connections=100
```
configmap.yaml
```yaml
   apiVersion: v1
   kind: ConfigMap
   metadata:
     name: mailu-config
     namespace: mailu-mailserver
   data:
    # Mailu main configuration file
    #
    # Most configuration variables can be modified through the Web interface,
    # these few settings must however be configured before starting the mail
    # server and require a restart upon change.

    ###################################
    # Common configuration variables
    ###################################

    # Set this to the path where Mailu data and configuration is stored
    ROOT: "/mailu"

    # Mailu version to run (1.0, 1.1, etc. or master)
    VERSION: "master"

    # Set to a randomly generated 16 bytes string
    SECRET_KEY: "1234567890abcdef"

    # Address where listening ports should bind
    BIND_ADDRESS4: "127.0.0.1"
    #BIND_ADDRESS6: "::1"

    # Main mail domain
    DOMAIN: "mydomain.com"

    # Hostnames for this server, separated with comas
    HOSTNAMES: "mail.mydomain.com"

    # Postmaster local part (will append the main mail domain)
    POSTMASTER: "admin"

    # Choose how secure connections will behave (value: letsencrypt, cert, notls, mail, mail-letsencrypt)
    TLS_FLAVOR: "cert"

    # Authentication rate limit (per source IP address)
    AUTH_RATELIMIT: "10/minute;1000/hour"

    # Opt-out of statistics, replace with "True" to opt out
    DISABLE_STATISTICS: "False"

    ###################################
    # Kubernetes configuration
    ###################################

    # Use Kubernetes Ingress Controller to handle all actions on port 80 and 443
    # This way we can make use of the advantages of the cert-manager deployment
    KUBERNETES_INGRESS: "false"

    # POD_ADDRESS_RANGE is normally provided by default with Kubernetes
    # Only use this value when you are using Flannel, Calico or a special kind of CNI
    # Provide the IPs of your network interface or bridge which is used for VXLAN network traffic
    POD_ADDRESS_RANGE: 10.233.0.0/16

    ###################################
    # Optional features
    ###################################

    # Expose the admin interface (value: true, false)
    ADMIN: "true"
    # Run the admin interface in debug mode
    #DEBUG: "True"

    # Choose which webmail to run if any (values: roundcube, rainloop, none)
    WEBMAIL: "roundcube"

    # Dav server implementation (value: radicale, none)
    WEBDAV: "radicale"

    # Antivirus solution (value: clamav, none)
    ANTIVIRUS: "clamav"

    ###################################
    # Mail settings
    ###################################

    # Message size limit in bytes
    # Default: accept messages up to 50MB
    MESSAGE_SIZE_LIMIT: "50000000"

    # Will relay all outgoing mails if configured
    #RELAYHOST=

    # This part is needed for the XCLIENT login for postfix. This should be the POD ADDRESS range
    # FRONT_ADDRESS: "front.mailu-mailserver.svc.cluster.local"
    FRONT_ADDRESS: 10.233.0.0/16

    # This value  is needed by the webmail to find the correct imap backend
    IMAP_ADDRESS: "imap.mailu-mailserver.svc.cluster.local"

    # This value is used by Dovecot to find the Redis server in the cluster
    REDIS_ADDRESS: "redis.mailu-mailserver.svc.cluster.local"

    # Fetchmail delay
    FETCHMAIL_DELAY: "600"

    # Recipient delimiter, character used to delimiter localpart from custom address part
    # e.g. localpart+custom@domain;tld
    RECIPIENT_DELIMITER: "+"

    # DMARC rua and ruf email
    DMARC_RUA: "root"
    DMARC_RUF: "root"

    # Welcome email, enable and set a topic and body if you wish to send welcome
    # emails to all users.
    WELCOME: "true"
    WELCOME_SUBJECT: "Welcome to your new email account"
    WELCOME_BODY: "Welcome to your new email account, if you can read this, then it is configured properly!"

    ###################################
    # Web settings
    ###################################

    # Path to the admin interface if enabled
    # Kubernetes addition: You need to change ALL the ingresses, when you want this URL to be different!!!
    WEB_ADMIN: "/admin"

    # Path to the webmail if enabled
    # Currently, this is not used, because we intended to use a different subdomain: webmail.example.com
    # This option can be added in a feature release
    WEB_WEBMAIL: "/webmail"

    # Website name
    SITENAME: "Mailu"

    # Linked Website URL
    WEBSITE: "https://mydomain.com"

    # Registration reCaptcha settings (warning, this has some privacy impact)
    # RECAPTCHA_PUBLIC_KEY=
    # RECAPTCHA_PRIVATE_KEY=

    # Domain registration, uncomment to enable
    # DOMAIN_REGISTRATION=true

    ###################################
    # Advanced settings
    ###################################

    # Create an admin account if it does not exist yet. It will also create the email domain for the account.
    # INITIAL_ADMIN_ACCOUNT: "admin"
    # INITIAL_ADMIN_DOMAIN: "example.com"
    # INITIAL_ADMIN_PW: "s3cr3t"

    # Docker-compose project name, this will prepended to containers names.
    COMPOSE_PROJECT_NAME: "mailu"

    # Default password scheme used for newly created accounts and changed passwords
    # (value: SHA512-CRYPT, SHA256-CRYPT, MD5-CRYPT, CRYPT)
    PASSWORD_SCHEME: "SHA512-CRYPT"

    # Header to take the real ip from
    # REAL_IP_HEADER: X-Real-IP

    # IPs for nginx set_real_ip_from (CIDR list separated by commas)
    # REAL_IP_FROM: 10.0.0.0/8,192.168.0.0/16

    # Host settings
    HOST_IMAP: "imap.mailu-mailserver.svc.cluster.local"
    HOST_POP3: "imap.mailu-mailserver.svc.cluster.local"
    HOST_SMTP: "smtp.mailu-mailserver.svc.cluster.local"
    HOST_AUTHSMTP: "smtp.mailu-mailserver.svc.cluster.local"
    HOST_WEBMAIL: "webmail.mailu-mailserver.svc.cluster.local"
    HOST_ADMIN: "admin.mailu-mailserver.svc.cluster.local"
    HOST_WEBDAV: "webdav.mailu-mailserver.svc.cluster.local:5232"
    HOST_ANTISPAM_MILTER: "antispam.mailu-mailserver.svc.cluster.local:11332"
    HOST_ANTISPAM_WEBUI: "antispam.mailu-mailserver.svc.cluster.local:11334"
    HOST_ANTIVIRUS: "antivirus.mailu-mailserver.svc.cluster.local:3310"
    HOST_REDIS: "redis.mailu-mailserver.svc.cluster.local"
```
traefik-ingress.yaml
```yaml
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: mailu-ingress-http
  namespace: mailu-mailserver
  labels:
    app: mailu
    role: mail
    tier: backend
spec:
  entryPoints:
    - web
  routes:
    - match: Host(`mail.mydomain.com`)
      kind: Rule
      services:
        - name: front
          port: 80
      middlewares:
        - name: https-redirectscheme
          namespace: traefik
---
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: mailu-ingress
  namespace: mailu-mailserver
  labels:
    app: mailu
    role: mail
    tier: backend
spec:
  entryPoints:
    - websecure
  routes:
    - match: Host(`mail.mydomain.com`)
      kind: Rule
      services:
        - name: front
          port: 80
      middlewares:
        - name: webmail-redirect
          namespace: mailu-mailserver
  tls:
    secretName: mailu-cert
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: webmail-redirect
  namespace: mailu-mailserver
spec:
  replacePathRegex:
    regex: ^\/?$ # match empty or with forward-slash
    replacement: /webmail
```

## Description
I was running Mailu for a couple weeks on my cluster but I was occasionally getting hang-ups where I couldn't authenticate. I dug into the logs for the front pod and found that I was getting blocked from my own server for going over the auth rate limit. Every single request - no matter who the request was coming from - was showing the client IP as the internal Kube IPs of the loadbalancer the traffic was coming from.

I figured I could have just bumped the auth rate limit up but that does not feel like a real solution. I had the Traefik service exposed with an external IP but it was not behind a load balancer so it never really got the real client IP to pass in. I figured I would add MetalLB so I could give the client IP over to Mailu and stop this madness. It doesn't seem to have made a difference - nginx front logs still show the client IP as internal IPs and my requests are still getting blocked half the time. I have a whoami service running in the same way and it does indeed show the correct client IP in the X-Real-Ip request header. I added that to the mailu configmap and restarted the pods but it still didn't help.

I also needed to add some new machines to the cluster so I was working on that at the same time. Once I got them added and the pods re-balanced out across the new machines all web traffic goes into an infinite redirect loop from the front nginx container. I am guessing what is happening is that it is sending the requests to the wrong node since the pods have moved around? I cannot explain this one away and can't seem to figure it out no matter what I try.

I have since removed MetalLB thinking it must have been the problem but I am still getting redirect loops even though it used to work with the same configuration.

In short - the only different in my cluster between when it was working and now is that there are two more nodes.

## Replication Steps
Either add MetalLB or add nodes to a bare metal cluster then delete Mailu pods. I am not 100% certain which triggered it.

## Expected behaviour
I expected passing in the real client IP and putting REAL_IP_HEADER and REAL_IP_FROM would mean Mailu would use the value of that header to identify clients. 
I expected to be able to scale nodes out and still access web admin and webmail.

## Logs
Front pod logs showing redirect loop:
```log
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
10.233.105.38 - - [05/Oct/2020:22:33:16 +0000] "GET /admin HTTP/1.1" 301 162 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:80.0) Gecko/20100101 Firefox/80.0"
```

Log showing auth rate limit exceeding - note the client IP:
```log
2020/10/05 22:33:19 [info] 10#10: *14782 client login failed: "Authentication rate limit from one source exceeded" while in http auth state, client: 10.233.70.45, server: 0.0.0.0:25, login: "me@mydomain.com"
```

**Comment by lub** [comment]
>  I expected to be able to scale nodes out and still access web admin and webmail.

This is something we are going to look at, but currently scaling by just running multiple containers with the same config (I'm don't know the k8s terminology for this, but I hope you understand what I mean) is not supported.
See https://github.com/Mailu/Mailu/blob/master/docs/swarm/master/README.md for some of the caveats. The document is mainly about swarm, but the scaling issues apply regardless of the orchestrator.

> I expected passing in the real client IP and putting REAL_IP_HEADER and REAL_IP_FROM would mean Mailu would use the value of that header to identify clients.

Do I understand correctly, that the logs you've posted were with both vars set? Because in your configmap.yaml they are commented out.

**Comment by lub** [comment]
Oh, it seems 'nodes' are just the kubernetes hosts itself. Nevermind the first part then.

Maybe someone with a multi-node kubernetes cluster can try to reproduce this?

**Comment by Matthew-Benson** [comment]
I  wanted to add a followup here. I just realized one difference between my configuration and the suggested configuration. I was using Traefik where I do not have the hostNetwork option enabled but I remembered the nginx lb for Mailu does have `hostNetwork: true` so perhaps that was the root cause of the problem? I don't have the same configuration anymore so this issue could be closed without being resolved unless someone else wants to take a deep dive into whatever happened here.
